events/controller.go

```diff
@@ -0,0 +1,63 @@
+package events
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/awslabs/operatorpkg/object"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/clock"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Controller[T client.Object] struct {
+	kind       string
+	startTime  time.Time
+	kubeClient client.Client
+}
+
+func NewController[T client.Object](client client.Client, clock clock.Clock) *Controller[T] {
+	return &Controller[T]{
+		kind:       reflect.TypeOf(object.New[T]()).Elem().Name(),
+		startTime:  clock.Now(),
+		kubeClient: client,
+	}
+}
+
+func (c *Controller[T]) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		For(&v1.Event{}, builder.WithPredicates(predicate.NewTypedPredicateFuncs(func(o client.Object) bool {
+			event := o.(*v1.Event)
+			return event.InvolvedObject.Kind == c.kind
+		}))).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
+		Named(fmt.Sprintf("operatorpkg.%s.events", strings.ToLower(c.kind))).
+		Complete(reconcile.AsReconciler(m.GetClient(), c))
+}
+
+func (c *Controller[T]) Reconcile(ctx context.Context, event *v1.Event) (reconcile.Result, error) {
+	// Deduplicate events on start-up
+	if c.startTime.Before(event.LastTimestamp.Time) {
+		group := strings.Split(event.InvolvedObject.APIVersion, "/")[0]
+		EventCount.Inc(map[string]string{
+			MetricLabelGroup:        group,
+			MetricLabelKind:         event.InvolvedObject.Kind,
+			MetricLabelNamespace:    event.InvolvedObject.Namespace,
```

Comment:
Let's drop namespace/name for now -- this is going to drastically increase our cardinality